### DISCUSSION TITLE
Nutzung mit Angular

### DISCUSSION THREAD

**Discussion by BenRichter** [public-ui/kolibri#858] [Q&A] [answered]
Hallo,
ich möchte public-ui in ein größeres Projekt einführen, dass jedoch noch eine Weile auf Angular 12 bleiben wird. Leider benötigen alle Versionen die via NPM verfügbar sind die 15.
Gibt es eine Möglichkeit trotzdem public-ui zu nutzen, ohne ein Umfangreiches Update durchzuführen?

**Comment by deleonio** [comment, marked as answer]
A: Wir können ja mal versuchen das Teil-Modul (adapter-angular) mit der Version 12 separat zu bauen.

B: Wir schauen was wir noch per peerDeps regeln können. (peerDeps: angular >= 12)

I prefer **B**.

**Reply by deleonio** [reply]
Hi @BenRichter,

bitte probiere mal v1.1.13-rc.3 aus - ob das Bauen mit Angular so möglich ist.

Wenn nicht, dann müssen wir ein separates ng12 Paket bauen.

VG @deleonio

**Reply by deleonio** [reply]
BTW @BenRichter: Falls Ihr ein eigenes Theme benötigt, können wir das hier diskutieren > https://github.com/public-ui/kolibri/discussions/894

**Reply by BenRichter** [reply]
Danke für die Schnelle Anpassung, jedoch fürchte ich wir sollten vorher wohl doch besser auf eine aktuelle Angular Version upgraden.

Beim einbinden von "KoliBriModule" in die imports der _app.module.ts_ entsteht ein Typescript Fehler in der Art für jede Komponente:

```PowerShell
Error: node_modules/@public-ui/angular/dist/components.d.ts:407:18 - error TS2314: Generic type 'ɵɵComponentDeclaration' requires 7 type argument(s).
                                                                                                                                                                                                                                                                                                                     
407     static ɵcmp: i0.ɵɵComponentDeclaration<KolToast, "kol-toast", never, { "_alert": "_alert"; "_hasCloser": "_hasCloser"; "_heading": "_heading"; "_level": "_level"; "_on": "_on"; "_show": "_show"; "_showDuration": "_showDuration"; "_type": "_type"; }, {}, never, ["*"], false, never>;
```

**Reply by deleonio** [reply]
👌 **Okay** @BenRichter 

🚀  dann versuchen wir es einfach mal mit einem dezidierten Angular ^12 Paket: [@public-ui/angular-v12](https://www.npmjs.com/package/@public-ui/angular-v12/v/1.1.13-rc.5?activeTab=versions)

Es wäre nett, wenn Ihr es ausprobiert, wir müssten uns dazu erst ein Projekt anlegen.

`package.json`
```json
{
  "dependencies": {
    "@public-ui/angular-v12": "1.1.13-rc.5"
  }
}
```

**Reply by BenRichter** [reply]
Vielen Dank! Es funktioniert nun mit den packages:
```
    "@public-ui/angular-v12": "^1.1.13-rc.5",
    "@public-ui/components": "^1.1.13-rc.7",
    "@public-ui/core": "^1.1.13-rc.7",
    "@public-ui/themes": "^1.1.13-rc.7",
```

Im app.module.ts werden die Komponenten wie folgt registriert:
```
import { KoliBriModule } from '@public-ui/angular-v12';
import { defineCustomElements } from '@public-ui/components/dist/loader';
import { register } from '@public-ui/core';
import { BMF } from '@public-ui/themes';

...

imports: [
  ...
  KoliBriModule, 
]

...

export class AppModule {
  constructor() {
    register(BMF, defineCustomElements);
  }
}
```

**Comment by deleonio** [comment]
Wir werden die Projekt-Hüllen für Angular 8, 9 und 10 wieder löschen - da darin offene CVE's sind.

<img width="679" alt="image" src="https://user-images.githubusercontent.com/6279703/204528155-291ac57c-eefd-47d3-a650-c8ff892d0ae5.png">
